faiss/impl/ProductAdditiveQuantizer.cpp

```diff
@@ -0,0 +1,358 @@
+
+            offset_m += q->M;
+            offset_d += q->d;
+        }
+    }
+}
+
+void ProductAdditiveQuantizer::compute_LUT(
+        size_t n,
+        const float* xq,
+        float* LUT,
+        float alpha,
+        long ld_lut) const {
+    // codebooks:  size (M * ksub, dsub)
+    // xq:         size (n, d)
+    // output LUT: size (n, M * ksub)
+
+    FINTEGER nqi = n;
+    // leading dimension of 'LUT' and 'xq'
+    FINTEGER ld_LUT = ld_lut > 0 ? ld_lut : total_codebook_size;
+    FINTEGER ld_xq = d;
+
+    float zero = 0;
+    size_t offset_d = 0;
+    size_t offset_cb = 0;
+    size_t offset_lut = 0;
+
+    for (size_t s = 0; s < nsplits; s++) {
+        const auto q = quantizers[s];
+
+        FINTEGER ncenti = q->total_codebook_size;
+        FINTEGER ld_cb = q->d; // leading dimension of 'codebooks'
+
+        auto codebooksi = codebooks.data() + offset_cb;
+        auto xqi = xq + offset_d;
+        auto LUTi = LUT + offset_lut;
+
+        sgemm_("Transposed",
+               "Not transposed",
+               &ncenti,
+               &nqi,
+               &ld_cb,
+               &alpha,
+               codebooksi,
+               &ld_cb,
+               xqi,
+               &ld_xq,
+               &zero,
+               LUTi,
+               &ld_LUT);
+
+        offset_d += q->d;
+        offset_cb += q->total_codebook_size * q->d;
+        offset_lut += q->total_codebook_size;
+    }
+}
+
+void ProductAdditiveQuantizer::set_verbose(bool verb) {
+    for (size_t s = 0; s < nsplits; s++) {
+        quantizers[s]->verbose = verb;
+    }
+    this->verbose = verb;
+}
+
+/*************************************
+ * Product Local Search Quantizer
+ ************************************/
+
+ProductLocalSearchQuantizer::ProductLocalSearchQuantizer(
+        size_t d,
+        size_t nsplits,
+        size_t Msub,
+        size_t nbits,
+        Search_type_t search_type) {
+    FAISS_THROW_IF_NOT(d % nsplits == 0);
+    size_t dsub = d / nsplits;
+    for (size_t i = 0; i < nsplits; i++) {
+        lsqs.emplace_back(dsub, Msub, nbits, ST_decompress);
+    }
+
+    std::vector<AdditiveQuantizer*> aqs;
+    for (size_t i = 0; i < nsplits; i++) {
+        aqs.push_back(&lsqs[i]);
+    }
+
+    init(d, aqs, search_type);
+}
+
+ProductLocalSearchQuantizer::ProductLocalSearchQuantizer() {}
+
+LocalSearchQuantizer* ProductLocalSearchQuantizer::subquantizer(size_t s) {
+    return &lsqs[s];
+}
+
+/*************************************
+ * Product Residual Quantizer
+ ************************************/
+
+ProductResidualQuantizer::ProductResidualQuantizer(
+        size_t d,
+        size_t nsplits,
+        size_t Msub,
+        size_t nbits,
+        Search_type_t search_type) {
+    FAISS_THROW_IF_NOT(d % nsplits == 0);
+    size_t dsub = d / nsplits;
+    std::vector<AdditiveQuantizer*> aqs;
+    for (size_t i = 0; i < nsplits; i++) {
+        rqs.emplace_back(dsub, Msub, nbits, ST_decompress);
+        aqs.push_back(&rqs[i]);
```

Comment:
BUG: the underlying memory of `rqs` could be changed during the loop. Use `reserve()` to fix it.